### ISSUE TITLE
Enabling overwriting and controlling provenance

### ISSUE THREAD

**Issue by cshenry** [ModelSEED/ModelSEED#69]
So I am growing increasingly confident we want to exercise more control over provenance. For example, in the current system if you want to add a media condition and run it in a model, you use the "addmedia" function, then you have to run "updateprovenance" on your model. You have to do both of these commands for every single media condition you added, which gets tedious, and also drives up storage space consumption.

For this reason, I've got the addmedia function doing an "overwrite" save instead of saving to a new UUID. It's a safe maneuver, because adding media is never likely to break a model or anything else, and it's not likely to be a change you'd want to revert.

Still, while filedb is working with the overwrites, I'm worried about how this is being done, and I'd like Paul to look over the changes and make sure it's okay. The overwites won't work at all in MongoDB at the moment. So we need our databases to work well with the overwrite saves, and we need to decide what our policy is going to be in the long term with object saving. I see the following options:

1.) We automatically keep provenance on any dangerous call (e.g. deletion or modification of subobjects), but we don't bother on safe calls (e.g. addition of subobjects) and object update
2.) We only create provenance objects when the user requests it by running a "create_provenance" command.

**Comment by pfrybar** [comment]
I'll have to figure out how to correctly perform an overwrite save for the FileDB. It should probably first delete the object, then save the new object, otherwise there will be orphaned data just sitting in the database which will not get cleaned up.

**Comment by cshenry** [comment]
Yeah, in fact, this is what is happening right now since I enabled the overwrite save in FileDB.
Although, honestly, a simple solution would be to not worry about the orphaned data and just periodically run a cleanup operation that rebuilds the file eliminate any data with no entries in the index...

**Comment by pfrybar** [comment]
Oops, clicked the wrong button, sorry

**Comment by pfrybar** [comment]
That's basically what I do for the FileDB (technically the KeyValueStore) right now. When you delete an object it just removes the reference, then every once in a while it will go through and create a new data file. To enable this however I think we need to delete the object before overwriting. I'll look into this to make sure.

I definitely agree that we need a better way to control the provenance of objects. Your example with the media conditions proves that. I'm not sure how we do this however.

1.) This might confuse the users (and the developers) as to which operations have been saved, and which have not. For example if they add a media condition, then want to remove it, can they simply run 'ms revert'? Also it might be difficult to determine which operations are "safe" and which are "dangerous".

2.) I feel like this might be the best approach. It has more of a 'git' feel to it, where you commit your changes you want to save. This might get a little complicated with linked objects however. And it would still be nice to be able to revert simple changes, like adding a media condition, without having to remember to save it every time.

Maybe we could do some sort of a hybrid system, where we save intermediate changes as well as larger commits. I'll have to think about this more.

**Comment by cshenry** [comment]
Okay, so here is my proposal:

1.) Users have to consciously save provenance on biochemistry and mapping when they are in the midst of altering these objects, but we will automatically save provenance on these objects when a new model is built or a new annotation is imported.

2.) We will save provenance on models and annotations every time... no need to the user to consciously do it.

3.) When you load a model that is linked to a provenance object that has been updated, the system automatically checks if it's provenance objects have descendants, and if so, it automatically updates to the latest provenance object (this would be nearly instantaneous, because it wouldn't load the provenance object until it has decided which one to load). Saved objects would be saved against the newest provenance, while the old object would still point to the original provenance object. We could have a global config that turns this feature on/off. If possible, we could have a global override on the commands that did the same. You would only turn this feature off if the auto update caused a problem for some reason.

This has the advantage of carefully maintaining all provenance, avoiding complexity in managing provenance, avoiding an explosion of new provenance objects, yet still enabling one to immediately run fba or gapfill against a media condition you just added to the database.
